spacy/pipeline/_parser_internals/_state.pxd

```diff
@@ -41,25 +40,32 @@ cdef cppclass StateC:
     int offset
     int _b_i
 
-    __init__(const TokenC* sent, int length) nogil:
+    __init__(const TokenC* sent, int length) nogil except +:
+        cdef int init_capacity
+
+        this._heads.resize(length)
+        this._unshiftable.resize(length)
+
+        # Reserve memory ahead of time to minimize allocations during parsing.
+        # The initial capacity set here ideally reflects the expected average-case/majority usage.
+        init_capacity = 20
+        this._stack.reserve(init_capacity)
+        this._rebuffer.reserve(init_capacity)
+        this._ents.reserve(init_capacity)
+        this._left_arcs.reserve(init_capacity)
+        this._right_arcs.reserve(init_capacity)
+        this.history.reserve(init_capacity)
+
         this._sent = sent
-        this._heads = <int*>calloc(length, sizeof(int))
-        if not (this._sent and this._heads):
-            with gil:
-                PyErr_SetFromErrno(MemoryError)
-                PyErr_CheckSignals()
         this.offset = 0
         this.length = length
         this._b_i = 0
         for i in range(length):
             this._heads[i] = -1
-            this._unshiftable.push_back(0)
+            this._unshiftable[i] = 0
```

Comment:
Doesn't `this._unshiftable.resize(length)` already set this to all-zeros? (Since the default constructor is used.)

Also, the resize for `_heads` could be changed to:

```
this._heads.resize(length, -1)
```

and then this loop could be eliminated completely.

Reply:
Indeed. Since the default size is zero, the `resize` call should default-insert all new values.